In [2]:
# Import the modules
import numpy as np
import pandas as pd
from pathlib import Path
from sklearn.metrics import balanced_accuracy_score, confusion_matrix, classification_report

---

## Split the Data into Training and Testing Sets

### Step 1: Read the `lending_data.csv` data from the `Resources` folder into a Pandas DataFrame.

In [3]:
# Read the CSV file from the Resources folder into a Pandas DataFrame
data_path = Path("Resources/lending_data.csv")
df = pd.read_csv(data_path)

# Review the DataFrame
print(df.head())

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  loan_status  
0                 1       22800            0  
1                 0       13600            0  
2                 0       16100            0  
3                 1       22700            0  
4                 1       23000            0  


### Step 2: Create the labels set (`y`)  from the “loan_status” column, and then create the features (`X`) DataFrame from the remaining columns.

In [4]:
# Separate the data into labels and features

# Separate the y variable, the labels
y = df['loan_status']

# Separate the X variable, the features
X = df.drop(columns=['loan_status'])


In [6]:
# Review the y variable Series
print(y.head())
# Review the distribution of the y variable 
print(y.value_counts())


0    0
1    0
2    0
3    0
4    0
Name: loan_status, dtype: int64
0    75036
1     2500
Name: loan_status, dtype: int64


In [7]:
# Review the first few rows of the X variable
print(X.head())

# Get a summary of the X variable
print(X.info())

   loan_size  interest_rate  borrower_income  debt_to_income  num_of_accounts  \
0    10700.0          7.672            52800        0.431818                5   
1     8400.0          6.692            43600        0.311927                3   
2     9000.0          6.963            46100        0.349241                3   
3    10700.0          7.664            52700        0.430740                5   
4    10800.0          7.698            53000        0.433962                5   

   derogatory_marks  total_debt  
0                 1       22800  
1                 0       13600  
2                 0       16100  
3                 1       22700  
4                 1       23000  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 77536 entries, 0 to 77535
Data columns (total 7 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   loan_size         77536 non-null  float64
 1   interest_rate     77536 non-null  float64
 2   borrower

### Step 3: Check the balance of the labels variable (`y`) by using the `value_counts` function.

In [8]:
# Check the balance of our target values
balance = y.value_counts()

print(balance)

0    75036
1     2500
Name: loan_status, dtype: int64


### Step 4: Split the data into training and testing datasets by using `train_test_split`.

In [9]:
# Import the train_test_learn module
from sklearn.model_selection import train_test_split

# Split the data using train_test_split
# Assign a random_state of 1 to the function
# Split the data using train_test_split with a random_state of 1
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)

#Print out the shapes of the training and testing datasets to verify the split
print("X_train shape:", X_train.shape)
print("X_test shape:", X_test.shape)
print("y_train shape:", y_train.shape)
print("y_test shape:", y_test.shape)

X_train shape: (58152, 7)
X_test shape: (19384, 7)
y_train shape: (58152,)
y_test shape: (19384,)


---

## Create a Logistic Regression Model with the Original Data

###  Step 1: Fit a logistic regression model by using the training data (`X_train` and `y_train`).

In [10]:
# Import the LogisticRegression module from SKLearn
from sklearn.linear_model import LogisticRegression

# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logistic_model = LogisticRegression(random_state=1)

# Fit the model using training data
logistic_model.fit(X_train, y_train)


LogisticRegression(random_state=1)

### Step 2: Save the predictions on the testing data labels by using the testing feature data (`X_test`) and the fitted model.

In [11]:
# Make a prediction using the testing data
y_pred = logistic_model.predict(X_test)
# Print out the first few predictions to verify
print(y_pred[:5])


[0 0 0 0 0]


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [12]:
# Print the balanced_accuracy score of the model
accuracy = balanced_accuracy_score(y_test, y_pred)
print(f"Balanced Accuracy Score: {accuracy:.4f}")

Balanced Accuracy Score: 0.9520


In [13]:
# Generate a confusion matrix for the model
confusion = confusion_matrix(y_test, y_pred)
print("Confusion Matrix:")
print(confusion)

Confusion Matrix:
[[18663   102]
 [   56   563]]


In [14]:
# Print the classification report for the model
report = classification_report(y_test, y_pred)
print("Classification Report:")
print(report)


Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.85      0.91      0.88       619

    accuracy                           0.99     19384
   macro avg       0.92      0.95      0.94     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question.

**Question:** How well does the logistic regression model predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

The logistic regression model predicts the 0 (healthy loan) labels with impressive accuracy. Specifically:

Precision for class 0 is 1.00 (or 100%), meaning that when the model predicts a loan is healthy, it's almost always correct.
Recall for class 0 is 0.99 (or 99%), suggesting that the model identifies 99% of the actual healthy loans.
The f1-score for class 0 is 1.00 (or 100%), indicating an excellent balance of precision and recall for predicting healthy loans.
For the 1 (high-risk loan) labels:

Precision is 0.85 (or 85%), meaning that when the model predicts a loan is high-risk, it's correct 85% of the time.
Recall is 0.91 (or 91%), suggesting that the model captures 91% of the actual high-risk loans.
The f1-score is 0.88 (or 88%), which is a strong indication of the model's balanced performance in identifying high-risk loans.
Overall, the logistic regression model does an excellent job of predicting both healthy and high-risk loans, with a total accuracy of 99% across all predictions. The model not only correctly identifies almost all of the healthy loans but also effectively flags a high percentage of the high-risk ones.

---

## Predict a Logistic Regression Model with Resampled Training Data

### Step 1: Use the `RandomOverSampler` module from the imbalanced-learn library to resample the data. Be sure to confirm that the labels have an equal number of data points. 

In [16]:
# Import the RandomOverSampler module form imbalanced-learn
from imblearn.over_sampling import RandomOverSampler

# Instantiate the random oversampler model
# # Assign a random_state parameter of 1 to the model
ros = RandomOverSampler(random_state=1)

# Fit the original training data to the random_oversampler model
X_resampled, y_resampled = ros.fit_resample(X_train, y_train)


0    56271
1    56271
Name: loan_status, dtype: int64


In [17]:
# Count the distinct values of the resampled labels data
print(y_resampled.value_counts())


0    56271
1    56271
Name: loan_status, dtype: int64


### Step 2: Use the `LogisticRegression` classifier and the resampled data to fit the model and make predictions.

In [18]:
# Instantiate the Logistic Regression model
# Assign a random_state parameter of 1 to the model
logreg_resampled = LogisticRegression(random_state=1)

# Fit the model using the resampled training data
logreg_resampled.fit(X_resampled, y_resampled)

# Make a prediction using the testing data
y_pred_resampled = logreg_resampled.predict(X_test)


### Step 3: Evaluate the model’s performance by doing the following:

* Calculate the accuracy score of the model.

* Generate a confusion matrix.

* Print the classification report.

In [19]:
# Print the balanced_accuracy score of the model 
balanced_accuracy_resampled = balanced_accuracy_score(y_test, y_pred_resampled)
print(f"Balanced Accuracy Score: {balanced_accuracy_resampled}")

Balanced Accuracy Score: 0.9936781215845847


In [20]:
# Generate a confusion matrix for the model
confusion_matrix_resampled = confusion_matrix(y_test, y_pred_resampled)
print("Confusion Matrix:")
print(confusion_matrix_resampled)

Confusion Matrix:
[[18649   116]
 [    4   615]]


In [21]:
# Print the classification report for the model
classification_report_resampled = classification_report(y_test, y_pred_resampled)
print("Classification Report:")
print(classification_report_resampled)

Classification Report:
              precision    recall  f1-score   support

           0       1.00      0.99      1.00     18765
           1       0.84      0.99      0.91       619

    accuracy                           0.99     19384
   macro avg       0.92      0.99      0.95     19384
weighted avg       0.99      0.99      0.99     19384



### Step 4: Answer the following question

**Question:** How well does the logistic regression model, fit with oversampled data, predict both the `0` (healthy loan) and `1` (high-risk loan) labels?

**Answer:** 
The logistic regression model, when fit with oversampled data, performs impressively well in predicting both the 0 (healthy loan) and 1 (high-risk loan) labels.

The balanced accuracy score of 0.9937 indicates that the model has a high accuracy in correctly predicting both classes, taking into account the imbalanced nature of the dataset.

The confusion matrix shows:

True Positives (TP) for healthy loans (0) as 18,649 and False Positives (FP) as 116, indicating a minor fraction of healthy loans were wrongly classified as high-risk.
True Positives (TP) for high-risk loans (1) as 615 and only 4 were classified wrongly as healthy loans. This is an impressive recall for high-risk loans.
The classification report provides a more detailed insight:

For healthy loans (0):
The precision is 1.00, indicating that almost all the loans the model identified as healthy were indeed healthy.
The recall is 0.99, suggesting that the model caught 99% of all the healthy loans.
For high-risk loans (1):
The precision is 0.84, meaning that of the loans the model identified as high-risk, 84% were indeed high-risk.
The recall is an impressive 0.99, indicating the model caught 99% of all the high-risk loans.
In summary, the logistic regression model trained on oversampled data exhibits a high level of accuracy and recall, especially for high-risk loans. This high recall is crucial for a lending institution since it indicates the model's ability to correctly identify the majority of high-risk loans, which could lead to potential defaults.